api/src/main/java/com/epam/pipeline/manager/log/LogManager.java

```diff
@@ -0,0 +1,269 @@
+import org.elasticsearch.search.builder.SearchSourceBuilder;
+import org.elasticsearch.search.sort.SortOrder;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.http.HttpStatus;
+import org.springframework.stereotype.Service;
+import org.springframework.util.Assert;
+import org.springframework.util.CollectionUtils;
+import org.springframework.util.StringUtils;
+
+import java.io.IOException;
+import java.time.LocalDateTime;
+import java.time.ZoneOffset;
+import java.time.format.DateTimeFormatter;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+@Slf4j
+@Service
+@Getter
+@Setter
+@RequiredArgsConstructor
+public class LogManager {
+
+    private static final IndicesOptions INDICES_OPTIONS = IndicesOptions.fromOptions(true,
+            SearchRequest.DEFAULT_INDICES_OPTIONS.allowNoIndices(),
+            SearchRequest.DEFAULT_INDICES_OPTIONS.expandWildcardsOpen(),
+            SearchRequest.DEFAULT_INDICES_OPTIONS.expandWildcardsClosed(),
+            SearchRequest.DEFAULT_INDICES_OPTIONS);
+
+    private static final String USER = "user";
+    private static final String TIMESTAMP = "@timestamp";
+    private static final String MESSAGE_TIMESTAMP = "message_timestamp";
+    private static final String HOSTNAME = "hostname";
+    private static final String SERVICE_NAME = "service_name";
+    private static final String TYPE = "type";
+    private static final String MESSAGE = "message";
+    private static final String SEVERITY = "level";
+    private static final DateTimeFormatter DATE_TIME_FORMATTER =
+            DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'");
+    private static final String DEFAULT_SEVERITY = "INFO";
+    public static final String ID = "_id";
+    public static final String SERVICE_ACCOUNT = "service_account";
+
+    private final GlobalSearchElasticHelper elasticHelper;
+    private final MessageHelper messageHelper;
+
+    @Value("${log.security.elastic.index.prefix}")
+    private String indexTemplate;
+
+    /**
+     * Searches log according to specified log filter.
+     * @param logFilter - filter for constructing elasticsearch query
+     * @return {@link LogPagination} object with related search result and additional information
+     * */
+    public LogPagination filter(final LogFilter logFilter) {
+        final LogPaginationRequest pagination = logFilter.getPagination();
+
+        Assert.notNull(pagination, messageHelper.getMessage(MessageConstants.ERROR_PAGINATION_IS_NOT_PROVIDED));
+        Assert.isTrue(pagination.getPageSize() > 0,
+                messageHelper.getMessage(MessageConstants.ERROR_INVALID_PAGE_INDEX_OR_SIZE));
+
+        final SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder()
+                .query(constructQueryFilter(logFilter))
+                .sort(MESSAGE_TIMESTAMP, SortOrder.ASC)
+                .sort(ID, SortOrder.ASC)
```

Comment:
As far as I remember sorting by id is not recommended due to performance considerations: "The _id field has a unique value per document but it is not recommended to use it as a tiebreaker directly" https://www.elastic.co/guide/en/elasticsearch/reference/6.8/search-request-search-after.html.  
I think sorting by timestamp shall be enough?

Reply:
Two fields is used here to have unique value from pair. Here I try to resolve problem when two messages can be acquired at the same time. But indeed we should get rid of using `_id` as  tiebreaker, now ingest pipeline creates second field `event_id` to use it instead of `_id`